### PR TITLE
Add Wiimote support

### PR DESCRIPTION
**Pull request by larsiusprime**
This adds wiimote support using either the MayFlash Dolphin Bar or the [HID-Wiimote driver](https://github.com/jloehr/HID-Wiimote).

Currently works and tested in flash and legacy. Does not currently work in next for some reason, but doesn't break anything, either. I think it just isn't detecting the gamepad ID for some reason.

Legacy support needs this (simple) pull request to lime to work: 
https://github.com/openfl/lime/pull/499

### CODE REVIEW

File: flixel/input/gamepad/FlxGamepad.hx [hunk 1]
```diff
@@ -86,40 +104,93 @@ class FlxGamepad implements IFlxDestroyable
...
 		_isChrome = (Capabilities.manufacturer == "Google Pepper");
 		#end
-		
-		pressed = new FlxGamepadButtonList(FlxInputState.PRESSED, this);
-		justPressed = new FlxGamepadButtonList(FlxInputState.JUST_PRESSED, this);
-		justReleased = new FlxGamepadButtonList(FlxInputState.JUST_RELEASED, this);
-		analog = new FlxGamepadAnalogList(this);
+	}
+	
+	public function traceAxes(f:FlxTimer):Void
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: These trace functions look like debug code?

File: flixel/input/gamepad/FlxGamepad.hx [hunk 2]
```diff
@@ -63,6 +73,14 @@ class FlxGamepad implements IFlxDestroyable
 	 * Helper class to get the justMoved, justReleased, and float values of analog input.
 	 */
 	public var analog(default, null):FlxGamepadAnalogList;
+	/**
+	 * Helper class to get the float values of motion-sensing input, if it is available
+	 */
+	public var motion(default, null):FlxGamepadMotionValueList;
+	/**
+	 * Helper class to get the float values of mouse-like pointer nput, if it is available
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: `nput`? :)

File: flixel/input/gamepad/FlxGamepad.hx [hunk 3]
```diff
@@ -33,6 +37,12 @@ class FlxGamepad implements IFlxDestroyable
 	public var model(default, set):FlxGamepadModel;
 	
 	/**
+	 * For gamepads that can have things plugged into them (the Wii Remote, basically).
+	 * Making the user set this helps
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: "helps..." what?

File: flixel/input/gamepad/FlxGamepad.hx [hunk 4]
```diff
@@ -719,13 +856,55 @@ class FlxGamepadAnalogStick
...
+	}
+	
+	public function toString():String
+	{
+		return("stick(" + x + "," + y + ",(" + rawUp + "," + rawDown + "," + rawLeft + "," + rawRight + ") @" + digitalThreshold + ":" + mode+")");
 	}
 }
 
+typedef FlxGamepadAnalogStickSettings = {
+	@:optional var up:Int;
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Same optional vs ? thing.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 3 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: haha, I did forget to finish that didn't I.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 4 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Does the ? go before the "var" or before the variable name?

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: There's no `var` with that syntax.

File: flixel/input/gamepad/FlxGamepadInputID.hx [hunk 5]
```diff
@@ -70,6 +70,25 @@ abstract FlxGamepadInputID(Int) from Int to Int
...
+	var TILT_PITCH           = 26;
+	/**tilting side-to-side (think "twisting", or "do a barrel roll!")**/
+	var TILT_ROLL            = 27;
+	
+	/**for a mouse-like input such as touch or IR camera. Horizontal axis.**/
+	var POINTER_X            = 28;
+	/**for a mouse-like input such as touch or IR camera. Vertical axis.**/
+	var POINTER_Y            = 29;
+	
+	/**an extra digital button that don't fit cleanly into the universal template**/
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadInputID.hx]: `don't fit` -> `doesn't fit`

File: flixel/input/gamepad/FlxGamepadManager.hx [hunk 6]
```diff
@@ -427,6 +427,8 @@ class FlxGamepadManager implements IFlxInputManager
 	
 	private function getModelFromFlashDeviceName(str:String):FlxGamepadModel
 	{
+		trace("str = " + str);
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadManager.hx]: Debug code.

File: flixel/input/gamepad/FlxGamepadManager.hx [hunk 7]
```diff
@@ -437,13 +439,20 @@ class FlxGamepadManager implements IFlxInputManager
 			}
 		}
 		
+		trace("str = " + str);
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadManager.hx]: Debug code.

File: flixel/input/gamepad/id/FlxGamepadMotionValueList.hx [hunk 8]
```diff
@@ -0,0 +1,29 @@
...
+ * A helper class for gamepad input.
+ * Provides optimized gamepad button checking using direct array access.
+ */
+@:keep
+class FlxGamepadMotionValueList
+{
+	private var gamepad:FlxGamepad;
+	
+	@:allow(flixel.input.gamepad.FlxGamepad)
+	public var available(default, null):Bool = true;
```
Comment (Gama11) [inline flixel/input/gamepad/id/FlxGamepadMotionValueList.hx]: Isn't this always true?

File: flixel/input/gamepad/id/LogitechID.hx [hunk 9]
```diff
@@ -7,6 +7,12 @@ import flixel.input.gamepad.FlxGamepad;
  */
 class LogitechID
 {
+	public static inline var SUPPORTS_MOTION = false;
```
Comment (Gama11) [inline flixel/input/gamepad/id/LogitechID.hx]: "Motion" means pointer input?

File: flixel/input/gamepad/id/FlxGamepadMotionValueList.hx (hunk 8 above)
Comment (larsiusprime) [inline flixel/input/gamepad/id/FlxGamepadMotionValueList.hx]: No, it gets set to false if the gamepad model is switched to something that doesn't support motion.

File: flixel/input/gamepad/FlxGamepadManager.hx [hunk 10]
```diff
@@ -500,16 +511,80 @@ class FlxGamepadManager implements IFlxInputManager
...
+					}
+				}
 			}
 		}
+		
+		gamepad.axis = newAxis;
+		
+		if (anyAbove)
+		{
+			//trace("axis = " + newAxis);
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadManager.hx]: Debug code.

File: flixel/input/gamepad/FlxGamepadManager.hx [hunk 11]
```diff
@@ -500,16 +511,80 @@ class FlxGamepadManager implements IFlxInputManager
...
+					}
+					else if (newVal >= neg && oldVal < neg)
+					{
+						     if (i == stick.x) digitalButton = stick.rawLeft;
+						else if (i == stick.y) digitalButton = stick.rawUp;
+						handleButtonUp(new JoystickEvent(JoystickEvent.BUTTON_UP, FlashEvent.bubbles, FlashEvent.cancelable, FlashEvent.device, digitalButton));
+					}
+					
+					//pressed/released for digital RIGHT/DOWN
+					if(newVal > pos && oldVal <= pos)
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadManager.hx]: `if(` -> `if (`

File: flixel/input/gamepad/FlxGamepadManager.hx [hunk 12]
```diff
@@ -500,16 +511,80 @@ class FlxGamepadManager implements IFlxInputManager
...
+				{
+					var newVal = newAxis[i];
+					var oldVal = oldAxis[i];
+					
+					var neg = stick.digitalThreshold * -1;
+					var pos = stick.digitalThreshold;
+					var digitalButton = -1;
+					
+					//pressed/released for digital LEFT/UP
+					if(newVal < neg && oldVal >= neg)
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadManager.hx]: `if(` -> `if (`

File: flixel/input/gamepad/FlxGamepadMapping.hx [hunk 13]
```diff
@@ -88,16 +130,48 @@ class FlxGamepadMapping
...
 	}
 	
+	/**
+	 * Returns 1 or -1 depending on whether this axis needs to be flipped
+	 * @param	AxisID
+	 * @return
+	 */
+	public function getFlipAxis(AxisID:Int):Int
+	{
+		return switch(model)
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadMapping.hx]: `switch(` -> `switch (`

File: flixel/input/gamepad/FlxGamepadMapping.hx [hunk 14]
```diff
@@ -88,16 +130,48 @@ class FlxGamepadMapping
...
+					}
 				default: null;
 			}
 		}
 		return null;
 	}
 	
+	/**
+	 * Returns 1 or -1 depending on whether this axis needs to be flipped
+	 * @param	AxisID
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepadMapping.hx]: Might as well remove the `@param` and `@return` here, there's no description for them.

File: flixel/input/gamepad/id/LogitechID.hx (hunk 9 above)
Comment (larsiusprime) [inline flixel/input/gamepad/id/LogitechID.hx]: Not quite. Motion and Pointer are two different things, at least based on how I've got it set up currently.

Motion is pitch/roll/yaw detection -- So far in all my tests the WiiMote supports pitch/roll, and the PS4/PS3 also support pitch/roll, but only the WiiMote reports them in a way I can reliably detect without crazy custom driver mappings.

The WiiMote also supports an Infrared camera which is what it uses to drive the on-screen pointer on the Wii/WiiU. Basically the infrared camera and the sensor bar combine to give you X/Y coordinates, and then you can multiply the (0.0-1.0),(0.01-1.0) X/Y values to get a pointer value.

The PS4 also has a touchpad, as does the Vita, the 3DS, the WiiU gamepad, and the Ouya, and so the way I thought of this, was that the WiiMote IR camera pointer has more in common with these touchpads than it does with "motion" as it just spits out analog X/Y coordinates, so we should have a separate "pointer" API for all of those.

As of right now I have only been able to get the WiiMote to be able to actually use the pointer API on the PC, but on the native console side all of the above could fill out the pointer API.

File: flixel/input/gamepad/FlxGamepad.hx [hunk 15]
```diff
@@ -719,13 +845,55 @@ class FlxGamepadAnalogStick
...
+	@:optional var down:Int;
+	@:optional var left:Int;
+	@:optional var right:Int;
+	@:optional var threshold:Float;
+	@:optional var mode:AnalogToDigitalMode;
+}
+
+enum AnalogToDigitalMode
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Missing Flx prefix.

File: flixel/input/gamepad/FlxGamepad.hx [hunk 16]
```diff
@@ -719,13 +845,55 @@ class FlxGamepadAnalogStick
...
+{
+	SendBoth;
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Does the "Send" add anything here?

File: flixel/input/gamepad/FlxGamepadManager.hx (hunk 7 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepadManager.hx]: On it.

File: flixel/input/gamepad/FlxGamepad.hx [hunk 17]
```diff
@@ -63,6 +73,14 @@ class FlxGamepad implements IFlxDestroyable
...
 	 */
 	public var analog(default, null):FlxGamepadAnalogList;
+	/**
+	 * Helper class to get the float values of motion-sensing input, if it is available
+	 */
+	public var motion(default, null):FlxGamepadMotionValueList;
+	/**
+	 * Helper class to get the float values of mouse-like pointer nput, if it is available
+	 */
+	public var pointer(default, null):FlxGamepadPointerValueList;
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Would it make more sense to use a `FlxPointer` for this instead? That's the base class for `FlxMouse` and `FlxTouch` -> consistent API.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 4 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: so just like this?

```haxe
typedef FlxGamepadAnalogStickSettings = {
	?up:Int;
	?down:Int;
	?left:Int;
	?right:Int;
	?threshold:Float;
	?mode:AnalogToDigitalMode;
}
```

File: flixel/input/gamepad/id/FlxGamepadPointerValueList.hx [hunk 18]
```diff
@@ -0,0 +1,29 @@
...
+ * A helper class for gamepad input.
+ * Provides optimized gamepad button checking using direct array access.
+ */
+@:keep
+class FlxGamepadPointerValueList
+{
+	private var gamepad:FlxGamepad;
+	
+	@:allow(flixel.input.gamepad.FlxGamepad)
+	public var available(default, null):Bool = true;
```
Comment (Gama11) [inline flixel/input/gamepad/id/FlxGamepadPointerValueList.hx]: Might make sense to change `available` to `isSupported`? That's what `FlxAccelerometer` uses -> consistent API.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 1 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: they are, but could be useful to keep around in the API if they're not actually called... maybe? I use them all the time.

File: flixel/input/gamepad/id/FlxGamepadPointerValueList.hx (hunk 18 above)
Comment (larsiusprime) [inline flixel/input/gamepad/id/FlxGamepadPointerValueList.hx]: sounds good to me.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 16 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Guess not.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 17 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Yeah probably! 

As a quick aside -- is it possible to override Flixel's mouse input with another source, so you could use a gamepad analog stick, or wiimote pointer input, or even keyboard presses, etc, to drive the mouse, rather than native mouse input? Would be super swell for accessibility.

Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Hmm, actually looking at it I'm not sure. As for right now I'm just spitting out the raw analog output from the controller, and there's a whole bunch of uses for it. I think ONE of those uses is definitely to map that to the screen coordinates (which seems to be FlxPointer's purpose), but the game designer / player could also just want the raw analog input for something else. 

I'd love the option to map it to the screen but I don't want to over-process the data before the user has access to it.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 15 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Fixed among other things.

**Comment by larsiusprime** [comment]
What else?

File: flixel/input/gamepad/FlxGamepad.hx (hunk 17 above)
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Wouldn't the values already be screen coordinates, and be mapped to world coordinates? It's not like the controller knows about flixel's coordinate system.

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Didn't you add setGlobalScreenPositionUnsafe() for having a different source?

Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: I did? Is that what the blame history shows? Hmmm.

In any case, the values are not actually screen coordinates, they come off of the controller as two analog values between 0 and 1. Not only do they have no knowledge of flixel's coordinate system, they have no knowledge of the world coordinate systems either. It's just two values between the minimum and maximum values of the x/y axes of the detection range, whether it's from the sensor bar or the touchpad.

I mean I suppose I could map it world (OpenFL stage) coordinates using FlxPointer. 

Basically what I'm asking is, do you think there's any value in also reporting the 0-1 values, or will users just always multiply it times screen coordinates and be really annoyed that it doesn't do it for them.

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: The blame history might not show that because FlxPointer didn't exist back then, but I'm sure you added it specifically for flixel-ui's custom cursor stuff.

That seems a bit misleading then? The usage of "pointer" led me to think that these coords are comparable to the x/y of the mouse / a touch. This at least needs to be documented properly - I don't think the [0, 1] range is mentioned anywhere either?

File: flixel/input/gamepad/FlxGamepad.hx (hunk 1 above)
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Meh, not sure I'm a fan of having that be part of the API... Perhaps we could have a  `FlxGamepad` [tracker profile](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/system/debug/Tracker.hx#L84) instead?

File: flixel/input/gamepad/FlxGamepad.hx [hunk 19]
```diff
@@ -719,13 +852,55 @@ class FlxGamepadAnalogStick
...
+			rawDown  = (settings.down  != null ? settings.down  : -1);
+			rawLeft  = (settings.left  != null ? settings.left  : -1);
+			rawRight = (settings.right != null ? settings.right : -1);
+			digitalThreshold = (settings.threshold != null ? settings.threshold : -1);
+		}
+	}
+	
+	public function toString():String
+	{
+		return("stick(" + x + "," + y + ",(" + rawUp + "," + rawDown + "," + rawLeft + "," + rawRight + ") @" + digitalThreshold + ":" + mode+")");
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Might make sense to use [Flixel's standard toString() logic](https://github.com/HaxeFlixel/flixel/blob/dev/flixel/FlxObject.hx#L1061-L1070)?

File: flixel/input/gamepad/FlxGamepad.hx (hunk 17 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Hmmm, this will take some thinking.

So, we don't have the consoles hooked up yet, but here's every device that might use the "gamepad.pointer" API:

- Nintendo 3DS
- Nintendo WiiU
- OUYA touchpad on controller
- PS4 touchpad on controller
- Steam controller (multiple touchpads)
- Wii remote (IR camera)

So the obvious super easy thing to do on the native console side is to rig up the input from the touchscreen to Lime/OpenFL's mouse/touch events, report absolute world coordinates, and call it a day. Flixel doesn't even need to "know" about this.

The problem with this solution is that you might have more than one device that has this kind of "pointery" input, and the mouse/touch API is built around the paradigm that there is ONE mouse/touch surface on the device/computer. So if we do that with the API, then you can't make a PS4 game where everyone is using their own touchpad.

On the 3DS obviously we should just use the same touch API we use for mobile. On the WiiU -- can you have more than one gamepad active on the same console?

So there's this issue to work out.

Other than that, what do you propose besides "gamead.pointer" as the name?

File: flixel/input/gamepad/FlxGamepad.hx (hunk 1 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: It takes five seconds to add them in when I need them, I'll just trash em.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 19 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: didn't know we had such a thing, thx.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 17 above)
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Pointer might be fine (unless you have a better name, I don't right know), as long as it's documented what the values mean.

Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Sure, let's go with that for now. Where should I put the documentation, in FlxGamepadPointerValueList or in FlxGamepad itself, right next to the pointer object declaration?

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Would make sense to have some general info on the field in FlxGamepad, and more concrete info on the x/y fields (value range, etc).

File: flixel/input/gamepad/FlxGamepad.hx [hunk 20]
```diff
@@ -86,40 +111,63 @@ class FlxGamepad implements IFlxDestroyable
...
+		
 		#if FLX_JOYSTICK_API
 		leftStick = getRawAnalogStick(FlxGamepadInputID.LEFT_ANALOG_STICK);
 		rightStick = getRawAnalogStick(FlxGamepadInputID.RIGHT_ANALOG_STICK);
 		#end
+		
 		return model;
 	}
 	
+	private var t:FlxTimer = null;
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Don't think this is used anywhere (anymore)?

File: flixel/input/gamepad/FlxGamepad.hx [hunk 21]
```diff
@@ -33,6 +37,18 @@ class FlxGamepad implements IFlxDestroyable
...
+	 * For gamepads that can have things plugged into them (the Wii Remote, basically).
+	 * Making the user set this helps Flixel properly interpret inputs properly. 
+	 * EX: if you plug a nunchuk into the Wii Remote, you will get different values for 
+	 * certain buttons than with the Wii Remote alone.
+	 * (This is probably why Wii games ask the player what control scheme they are using.)
+	 * 
+	 * In the future, this could also be used for any attachment that exposes new API features
+	 * to the controller, e.g. a microphone or headset
+	 */
+	public var attachment(default, set):FlxGamepadModelAttachment;
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Hm... in theory, a gamepad could have multiple attachements?

File: flixel/input/gamepad/FlxGamepad.hx [hunk 22]
```diff
@@ -719,11 +822,61 @@ class FlxGamepadAnalogStick
...
 	public var y(default, null):Int;
 	
-	public function new(x:Int, y:Int)
+	//these values let the analog stick to send digital inputs to, say, the dpad
+	public var rawUp(default, null):Int = -1;
+	public var rawDown(default, null):Int = -1;
+	public var rawLeft(default, null):Int = -1;
+	public var rawRight(default, null):Int = -1;
+	
+	//the value the dpad must be above before digital inputs are sent
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: If you want this to appear in completion, it needs to be a `/**` comment I think.

File: flixel/input/gamepad/FlxGamepad.hx [hunk 23]
```diff
@@ -719,11 +822,61 @@ class FlxGamepadAnalogStick
...
+	public var digitalThreshold(default, null):Float = 0.5;
+	
+	//when analog inputs are received, how to process them digitally
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Same here.

File: flixel/input/gamepad/FlxGamepad.hx [hunk 24]
```diff
@@ -154,6 +203,11 @@ class FlxGamepad implements IFlxDestroyable
 		return gamepadButton;
 	}
 	
+	public inline function getFlipAxis(AxisID:Int):Int
```
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Could this be used to invert the y axis of the Xbox 360 controller in chrome, which is currently hardcoded in getYAxis?

File: flixel/input/gamepad/FlxGamepad.hx (hunk 21 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: In theory, yes; but let's wait until we actually have a use case for that ... [YAGNI](https://en.wikipedia.org/wiki/You_aren%27t_gonna_need_it) and all that.

Unless you can think of something super obvious that we could test with.

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Yeah, this was more of a general question rather than "this needs to be done differently".

File: flixel/input/gamepad/FlxGamepad.hx (hunk 24 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: I originally added it to fix a bug that has since been fixed by other means, but in short, YES. The idea was to move as much weird hardware-and-platform specific fixup code and mapping logic to the SomeControllerID.hx classes and out of the FlxGamepadMapping class.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 21 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Sure. But I'm also super interested to know if you have a game controller like that in mind; I'm already thinking about other stuff we could do with attachments -- X360 has the microphone/headset for instance, but not sure if we have any chance of exposing API's for that.

Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: No, wasn't thinking about any controller in particular.

File: flixel/input/gamepad/FlxGamepad.hx (hunk 24 above)
Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Good catch on this, as it was also broken on "Adobe Windows" Capabilities.manufacturer value (ie, the standalone flash player on windows).

Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: This fix is in now.

Comment (larsiusprime) [inline flixel/input/gamepad/FlxGamepad.hx]: Okay maybe I'm crazy, but on my computer chrome doesn't flip the Y axis in flash (so manually inverting it messes it up)... only on the standalone windows player do I have to flip it. 

It even says manufacturer = "Google Pepper" in the log trace in flash/chrome, so I dunno...

I'll commit what works here and you tell me what's happening for you.

**Comment by larsiusprime** [comment]
How bout now?

**Comment by Gama11** [comment]
HTML5 builds are failing (see travis).

File: flixel/input/gamepad/FlxGamepad.hx (hunk 24 above)
Comment (Gama11) [inline flixel/input/gamepad/FlxGamepad.hx]: Ah, right, I didn't remember that correctly. That's what the old code did as well actually.

```haxe
if (model == XBox360 && !_isChrome)
	axisValue = -axisValue;
```

**Comment by Gama11** [comment]
I can't test if the whole Wiimote stuff actually works, but the API looks ok to me now (and the Xbox 360 stuff still works). Thanks! :+1:

**Comment by larsiusprime** [comment]
Does the fact that we test HTML5 in travis imply that flixel works in HTML5? If so I can hook up the gamepad ID's for that too... are there any special instructions? (I just tried and got a black screen).

**Comment by Gama11** [comment]
That alone just implies that it compiles.. but, yes, in general it should also work. Mode works fine for example.

Getting any errors in the js console? What browser? Browsers other than Chrome have some issues atm I think.

**Comment by larsiusprime** [comment]
Cool, that's a good starting point, I'll go from there. Are we using canvas or GL, and/or what custom backend/add'l dependencies, if any?

**Comment by Gama11** [comment]
You shouldn't need to add anything. I think OpenFL ises canvas by default, haven't tested GL.

I'm not sure Lime's gamepad API is even implemented on HTML5 right now?

**Comment by sruloart** [comment]
BTW, If the black screen issue continues see [here](https://github.com/openfl/lime/issues/493#issuecomment-120238001) and [here](https://github.com/HaxeFlixel/flixel/issues/1462#issuecomment-120654596). 

(and AFAIK for dev it's canvas + Openfl (legacy) html5 backend).

**Comment by larsiusprime** [comment]
This should help with setting up gamepads for HTML5, I think:
http://www.html5rocks.com/en/tutorials/doodles/gamepad/gamepad-tester/tester.html

**Comment by sruloart** [comment]
It's only implemented in bitfive, looks like, not openfl...
